core/server/master/src/main/java/alluxio/master/file/InodeSyncStream.java

```diff
@@ -0,0 +1,810 @@
+   * @param loadOnly whether to only load new metadata, rather than update existing metadata
+   */
+  public InodeSyncStream(LockedInodePath rootPath, ExecutorService concurrencyService,
+      DefaultFileSystemMaster fsMaster, InodeTree inodeTree, ReadOnlyInodeStore inodeStore,
+      InodeLockManager inodeLockManager, MountTable mountTable, RpcContext rpcContext,
+      DescendantType descendantType, UfsSyncPathCache ufsSyncPathCache,
+      FileSystemMasterCommonPOptions options, boolean isGetFileInfo, boolean forceSync,
+      boolean loadOnly) {
+    mDescendantType = descendantType;
+    mFsMaster = fsMaster;
+    mSyncMetadataQ = new ConcurrentLinkedQueue<>();
+    mInodeLockManager = inodeLockManager;
+    mInodeStore = inodeStore;
+    mInodeTree = inodeTree;
+    mMountTable = mountTable;
+    mRpcContext = rpcContext;
+    mStatusCache = new UfsStatusCache(fsMaster.mSyncPrefetchExecutor);
+    mUfsSyncPathCache = ufsSyncPathCache;
+    mShouldSync = forceSync;
+    mRootPath = rootPath;
+    mSyncOptions = options;
+    mIsGetFileInfo = isGetFileInfo;
+    mLoadOnly = loadOnly;
+    mSyncPathJobs = new LinkedList<>();
+    mMetadataSyncService = concurrencyService;
+  }
+
+  /**
+   * Sync the metadata according the the root path the stream was created with.
+   *
+   * @return true if at least one path was synced
+   */
+  public boolean sync() {
+    // The high-level process for the syncing is:
+    // 1. Given an Alluxio path, determine if it is not consistent with the corresponding UFS path.
+    //     this means the UFS path does not exist, or has metadata which differs from Alluxio
+    // 2. If only the metadata changed, update the inode with the new metadata
+    // 3. If the path does not exist in the UFS, delete the inode in Alluxio
+    // 4. If not deleted, load metadata from the UFS
+    // 5. If a recursive sync, add children inodes to sync queue
+    int syncPathCount = 0;
+    int stopNum = -1; // stop syncing when we've processed this many paths. -1 for infinite
+
+    try {
+      syncInodeMetadata(mRootPath);
+      syncPathCount++;
+      if (mDescendantType == DescendantType.ONE) {
+        // If descendantType is ONE, then we shouldn't process any more paths except for those
+        // currently in the queue
+        stopNum = mSyncMetadataQ.size();
+      }
+
+      // process the sync result for the original path
+      try {
+        mRootPath.traverse();
+      } catch (InvalidPathException e) {
+        throw new RuntimeException(e);
+      }
+    } catch (AccessControlException | BlockInfoException | FileAlreadyCompletedException
+        | FileDoesNotExistException | InvalidFileSizeException | InvalidPathException
+        | IOException e) {
+      LOG.warn("FAILED TO SYNC METADATA: {}", e.getMessage(), e);
+    } finally {
+      // regardless of the outcome, remove the UfsStatus for this path from the cache
+      mStatusCache.remove(mRootPath.getUri());
```

Comment:
why remove the status?

Reply:
This will drop all references to the statuses allowing the JVM to GC it. If we don't remove it then for a huge namespace the statusCache can become large